vimish-fold.el

```diff
@@ -224,35 +240,91 @@ This includes fringe bitmaps and faces."
         '(vimish-fold--folded
           vimish-fold--unfolded)))
 
+(defun vimish-fold--folded-vimish-overlay-p (overlay)
+  "Detect if given OVERLAY is created by this package and unfolded."
+  (memq (overlay-get overlay 'type) '(vimish-fold--folded)))
+
+(defun vimish-fold--visible-vimish-overlay-p (overlay)
+  "Detect if given OVERLAY is visible or not."
+  (memq (overlay-get overlay 'visibility) '(vimish-fold--visible)))
+
+(defun vimish-fold--nested-vimish-overlay-p (overlay)
+  "Detect if given OVERLAY is nested or not."
+  (if (overlay-get overlay 'nested)
+      t
+    nil))
+
+(defun vimish-fold--overlay-min-start (ov1 ov2)
+  "Return the overlay between OV1 and OV2 that has the min `overlay-start'."
+  (if (< (overlay-start ov1) (overlay-start ov2))
+      t
+    nil))
+
+(defun vimish-fold--overlay-max-start (ov1 ov2)
+  "Return the overlay between OV1 and OV2 that has the max `overlay-start'."
+  (if (> (overlay-start ov1) (overlay-start ov2))
+      ov1
+    ov2))
+
 ;;;###autoload
-(defun vimish-fold (beg end)
-  "Fold active region staring at BEG, ending at END."
+(defun vimish-fold (beg end &optional nested)
+  "Fold active region starting at BEG, ending at END.
```

Comment:
As a user, why should I care about manually telling `vimish-fold` whether my new fold happens to be nested in some other fold or not? As I understand it, the fact whether the new fold is nested or not is determined entirely by position of the new fold and existing folds in the buffer.